Demo/MyPlayground.playground/Pages/AdvancePractice.xcplaygroundpage/Contents.swift

```diff
@@ -0,0 +1,671 @@
+
+
+class Book5 {
+    var name: String = "" {
+        willSet {
+            print("[willSet] newValue:", newValue)
+        }
+        didSet {
+            print("[didSet] oldValue:", oldValue)
+        }
+    }
+}
+let book51 = Book5()
+book51.name = book1.name
+print(book51)
+
+func change(_ e: inout String) {
+    e = "Changed value"
+}
+change(&book51.name)
+
+
+/**
+ Propertt wrapper
+ */
+
+@propertyWrapper
+struct EvenOnly {
+    private var num = 0
+    var wrappedValue: Int {
+        get { num }
+        set {
+            if(newValue%2 != 0) {
+                print("can't use odd numbers")
+                return
+            }
+            num = newValue
+        }
+    }
+}
+
+class Lift {
+    @EvenOnly var floors: Int
+}
+
+let lift = Lift()
+lift.floors = 6
+lift.floors = 5
+print(lift.floors)
+
+/**
+@propertyWrapper
+class PathFilter {
+    private let illigleChars: Set<Character> = ["\\", "/", ":", "*", "?", "\"", "<", ">", "|"]
+    private var name: String
+    
+    var wrappedValue: String {
+        get { return name }
+        set {
+            newValue.forEach {
+                if illigleChars.contains($0) {
+                    print("invalid file name!")
+                    return
+                }
+            }
+            name = newValue
+        }
+    }
+    
+    init() {
+        name = ""
+    }
+    
+    init(wrappedValue: String) {
+        wrappedValue.forEach { //'self' captured by a closure before all members were initialized
+            if illigleChars.contains($0) {
+                print("invalid file name!")
+                return
+            }
+        }
+        name = wrappedValue
+    }
+}
+
+struct FileProvider {
+    @PathFilter var name: String = ""
+}
+var fp = FileProvider(name: "hello")
+print(fp.name)
+*/
+
+class Counter {
+    var count: Int
+    
+    init() {
+        count = 0
+    }
+    
+    init(initial: Int) {
+        count = initial
+    }
+    
+    func inc(by: Int = 1) {
+        count += by
+    }
+    
+    func dec(by: Int = 1) {
+        count -= by
+    }
+}
+
+let counter = Counter(initial: 5)
+counter.inc()
+counter.dec(by: 3)
+print(counter.count)
+
+
+/**
+ Mutating
+ */
+
+struct Counter2 {
+    var count = 0
+    
+    mutating func inc(by: Int = 1) {
+        count += by
+    }
+    
+    mutating func dec(by: Int = 1) {
+        count -= by
+    }
+    
+    mutating func reset() {
+        self = Counter2()
+    }
+}
+
+var counter2 = Counter2()
+counter2.inc(by: 50)
+print(counter2.count)
+counter2.reset()
+print(counter2.count)
+
+enum Signal: String {
+    case green, yellow, red
+    
+    mutating func next() -> Signal {
+        switch self {
+        case .green:
+            self = .yellow
+        case .yellow:
+            self = .red
+        case .red:
+            self = .green
+        }
+        return self
+    }
+}
+var signal = Signal(rawValue: "green")
+print(signal!.rawValue)
```

Comment:
Avoid using "!" use nil coalescing "??" by providing default value

Reply:
Implemented changes in cb1362cde6d24923d5b49f390ba5c8e3319099f8